[4기 박은영] TodoList with Ajax

eyabc opened the pull request:
음일단 ! step2 를 마무리 했습니다.
```shell script
.
├── LICENSE
├── README.md
├── index.html
├── js-todo-list-step2.iml
└── src
    ├── app.js
    ├── constants
    │   ├── filter.js
    │   ├── index.js
    │   └── messageAPI.js
    ├── core
    │   └── state.js
    ├── css
    │   └── app.css
    ├── endpoint
    │   ├── HttpMethod.js
    │   ├── RestApi.js
    │   ├── api
    │   │   └── user.js
    │   ├── facade.js
    │   └── service.js
    ├── env.js
    ├── eventHandler.js
    ├── store
    │   └── index.js
    ├── template
    │   ├── Todo
    │   │   ├── TodoContainer.js
    │   │   ├── TodoCount.js
    │   │   ├── TodoCountContainer.js
    │   │   ├── TodoFilter.js
    │   │   ├── TodoInput.js
    │   │   ├── TodoItem.js
    │   │   ├── TodoLabel.js
    │   │   ├── TodoList.js
    │   │   └── TodoLoading.js
    │   ├── User
    │   │   ├── UserList.js
    │   │   ├── UserListContainer.js
    │   │   └── UserTitle.js
    │   └── index.js
    ├── utils.js
    └── validator.js
```

## step3 에서 공부해 보고 싶은것 
1.  상태관리에 subscribe 를 직접하지 않아도, 상태를 사용함으로써 최초 한번 subscribe 를 할 수 있는 행위
2. 컴포넌트 내 state 와, 전역 state 를 관리하는 방법에 대한 공부. 
3. 효율적인 렌더링을 위한 props diff 계산법이나 알고리즘
4. 가상 DOM 라이브러리 snabbdom 등을 조합
4. Flux 아키텍처 안티 패턴 
  - Q: 하위 component 마다 props 로 내려받은 item id 로 store 에서 필요한 값 탐색
  - 컴포넌트 재활용이 힘들다. 
  - 다수의 하위 컴포넌트들이 계산을 하면서 컴포넌트의 개수에 따라 연산이 중복되기에 탐색시간이 N배 증가하게 됩니다.
  - Q: 데이터를 obj 으로 만들면 탐색시간이 덜 걸릴 것 같음
     - 탐색 시간은 데이터 타입보다는 데이터를 어디에서 어떻게 다룰까가 초점입니다!
     
5. 데이터를 부모에서 순수한 props를 하위 컴포넌트 내려주기
    - 계산된 데이터를 다른 하위 컴포넌트에서 재활용할 할 수 있다.
6. 최적화 부분은 탐색시간이나 계산 값을 캐싱한다거나 메모제이션하는 기법을 주로 활용합니다.
7. api 부분을 역할을 더 잘 나누고 통일성 만들어보기. 이름도 잘 지어보기
8. step2 에서 최종적으로 todo 하나의 우선순위를 수정하거나 삭제하는 등 동작을 했을 때, 모든 todoList 를 재렌더링 하는 방법으로 하였지만.
매우 느려졌다. 문제점을 찾아서 해결해야 한다.
9. dom object 를 쓰면서 성능에 대해 신경쓰고 mdn 도 잘 읽어보기


https://meetup.toast.com/posts/158
https://www.youtube.com/watch?time_continue=7&v=o4meZ7MRd5o&feature=emb_title

7. 스텝 2는 id 별로 비공용 todoList 를 쓴다고 가정하고 매번 todoList 를 업데이트 하지 않는것으로 바꾸기

## step2 에서 고민해 본 것
1. 모든 요소를 컴포넌트로 만들기


```
## 🎯 요구사항

- [x] 1. User 추가하기
- [x] 2. User의 투두리스트 불러오기
- [x] 3. User 삭제하기
- [x] 3. todoItem 추가하기
- [x] 4. todoItem 불러오기
- [x] 5. todoItem complete하기
- [x] 6. todoItem 삭제하기
- [x] 7. todoItem contents 내용 수정하기


## 🎯🎯 심화 요구사항

- [x] 1. 데이터를 불러오기전 로딩바를 이용해, 사용자가 데이터가 불러와지고 있다는 것을 보여줍니다.
- [x] 2. fetch api 사용하는 부분을 async await을 사용하여 리팩토링합니다.
- [x] 3. github issue에서 라벨을 붙이는 것처럼, 우선순위에 따라서 badge를 추가합니다.
- [x] 4. ES6 impot & export를 이용해 자바스크립트 파일을 리팩토링합니다.

## 🕵️‍♂️ 제약사항

- [x] 1. User의 이름은 최소 2글자 이상이어야 합니다.

eastjun-dev committed:
Update README.md

eastjun-dev committed:
Update README.md

eyabc committed:
feat: init 컴포넌트 분리

eyabc committed:
feat: api method 추가
RestApi.js, HttpMethod.js 는 준일님 코드를 참고하였습니다.

eyabc committed:
refactor: endpoint 매개변수 및 모듈 변경

eyabc committed:
feat: 이벤트 등록 모듈,

eyabc committed:
feat: User 등록 api 호출

eyabc committed:
refactor: ripple 스타일 수정

eyabc committed:
feat: userList 를 가져오는 api 실행 및 렌더

eyabc committed:
feat: 첫 렌더시, userList 의 첫번째 사람이 선택됨

eyabc committed:
feat: 첫 렌더시 User title 변경

eyabc committed:
fix: user 추가시, name 이 없을때 버그 수정

eyabc committed:
refactor: 정리

eyabc committed:
feat: user 를 추가하면 userList 가 다시 렌더링

eyabc committed:
feat: user 를 추가하면 추가한 user 가 선택됨

eyabc committed:
refactor: 이벤트와 이벤트 콜백 파일 분리, api 를 호출하는 함수를 모은 service 모듈 생성

eyabc committed:
feat: user 를 변경할 때, title 도 변경.

eyabc committed:
feat: todoList template 추가

eyabc committed:
feat: user 가 바뀌었을 때 todoList 도 변경

eyabc committed:
feat: userList 와 userTodo 아이템이 로드 될 때 loading ui 를 보여줌

eyabc committed:
observer 를 쓰기 위해 컴포넌트 리팩토링 및, 유저 삭제 구현

eyabc committed:
fix: user 가 있을때만 user 를 삭제할 수 있음

eyabc committed:
feat: user Item 추가

eyabc committed:
feat: user 가 없을 때 UX

eyabc committed:
feat:
1. user item complete toggle
2. user item delete

eyabc committed:
refactor: 유저 item 추가 에러처리 변경

eyabc committed:
fix: 유저 item 추가가 두번되는 버그 수정

eyabc committed:
feat: user 아이템 변경

eyabc committed:
feat: user 아이템 우선순위 변경, 유저 아이템 모두 삭제 service 추가

eyabc committed:
fix: user 아이템 edit mode 클래스 추가

eyabc committed:
refactor: TodoLabel 도 return { dom, render } 구조로 변경

eyabc committed:
feat: item 우선순위 변경

eyabc committed:
fix: contents 내용 버그 수정

eyabc committed:
feat: user todoList 모두 삭제

eyabc committed:
fix: user todoList 불필요한 변수 삭제

eyabc committed:
feat: filter 선택

eyabc committed:
feat: complete toggle 을 했을때 todolist render

eyabc committed:
feat: Count 렌더

eyabc committed:
fix: Count 버그 수정

eyabc committed:
fix: 모두 삭제 버그 수정

eyabc committed:
feat: api 구조 변경

eyabc committed:
refactor: 에러 처리 변경, 서비스 이름 변경

eyabc committed:
refactor: baseURL set 변경

eyabc committed:
refactor: loading bar & dom root 파일 위치 수정

eyabc committed:
fix: complete toggle 버그 수정

eyabc committed:
refactor: template 구조 수정

eyabc committed:
refactor: event handler 1차분리, dom -> createDOM 메서드 생성

JunilHwang commented on src/app.js:
```diff
@@ -0,0 +1,10 @@
+
+const $app = document.getElementById('app');
+
+(async () => await initStore())()
+  .then(() => Home($app));
```
```suggestion
initStore().then(() => Home($app));
```
이렇게 해도 된답니다! 굳이 async/await + 즉시실행 함수까지 할 필요는 없을 것 같아요~

CianJS replied:
혹시 이렇게하면 그저 가독성이 좋아지는 것인지 아니면 성능면에서 어떤 차이가 나는 것인지 궁금합니다! @JunilHwang

JunilHwang commented on src/constants/index.js:
```diff
@@ -0,0 +1,7 @@
+import { filter } from './filter.js';
+
+export const ALL  = 'all';
+export const ACTIVE  = 'active';
+export const COMPLETED  = 'completed';
+
+export { filter };
```
```suggestion
export * from './filter.js';
export const ALL  = 'all';
export const ACTIVE  = 'active';
export const COMPLETED  = 'completed';
```
이렇게 표현할 수 있습니다

eyabc replied:
https://github.com/airbnb/javascript#modules--no-export-from-import

airbnb module 의 스타일 가이드 10.6 를 읽어보면서 분리 하는 것이 좋다고 하여 분리 해보았습니다
```
from 으로 Import 후 직접적인 export X
간결하지만, export / import 의 일관성을 위해 분리한다.
```

JunilHwang commented on src/constants/messageAPI.js:
```diff
@@ -0,0 +1,14 @@
+const ERROR = {
+  NO_USER: '해당 이름의 user가 없습니다.',  // used: createUserTodoItem
+  NO_USER2: '해당 user를 찾는데 에러가 발생했습니다.', // readUserTodoItems
+  NO_USER3: '해당 유저가 존재하지 않습니다.', // updateUserTodoItemComplete, deleteUserItemService, putUserItemService, deleteUserItemsAllService, updateUserTodoItemPriority
+  UPDATE_TODO_ITEM: 'Todo Item을 수정하는데 에러가 발생했습니다.', // updateUserTodoItemComplete, putUserItemService, updateUserTodoItemPriority
+  DELETE_TODO_ITEM: 'Todo Item을 삭제하는데 에러가 발생했습니다.',// deleteUserItemService
+};
+
+const SUCCESS = {
+  DELETE_TODO_ITEM_ALL: '해당 유저의 Todo Item 을 모두 삭제하였습니다.'
+};
+
+export { ERROR, SUCCESS };
+
```
```suggestion
export const ERROR = {
  NO_USER: '해당 이름의 user가 없습니다.',  // used: createUserTodoItem
  NO_USER2: '해당 user를 찾는데 에러가 발생했습니다.', // readUserTodoItems
  NO_USER3: '해당 유저가 존재하지 않습니다.', // updateUserTodoItemComplete, deleteUserItemService, putUserItemService, deleteUserItemsAllService, updateUserTodoItemPriority
  UPDATE_TODO_ITEM: 'Todo Item을 수정하는데 에러가 발생했습니다.', // updateUserTodoItemComplete, putUserItemService, updateUserTodoItemPriority
  DELETE_TODO_ITEM: 'Todo Item을 삭제하는데 에러가 발생했습니다.',// deleteUserItemService
};

export const SUCCESS = {
  DELETE_TODO_ITEM_ALL: '해당 유저의 Todo Item 을 모두 삭제하였습니다.'
};

```

그냥 바로 export 해주세요!

JunilHwang commented on src/core/state.js:
```diff
@@ -0,0 +1,19 @@
+  let _state = initVal;
+  const observable = new Set();
+
+  const state = (render) => {
+    if (render) observable.add(render);
+    return _state;
+  };
+
+  const setState = (newValue) => {
+      _state = newValue;
+      render();
+  };
+
```
심플하게 잘 구성하셨네요! 👍

JunilHwang commented on src/core/state.js:
```diff
@@ -0,0 +1,19 @@
+export const useState = (initVal = undefined) => {
+  let _state = initVal;
+  const observable = new Set();
```
이 변수는 아예 components 라고 하는게 어떨까요? 혹은 observers가 더 어울리는 것 같아요~

eyabc replied:
1. Observer : Observable 을 보고 행동을 알리는 주체
2. Observable : Observer 에 의하여 행동을 하는 것들

이렇게 이해 하고 있는데 Observable 에는 render 라는 행동의 대상들이 들어가 있으므로 Observable 가 문맥상으로 더 어울리지 않을까요?

JunilHwang commented on src/endpoint/RestApi.js:
```diff
@@ -0,0 +1,45 @@
+
+  let id = -1;
+  const race = await Promise.race([
+    new Promise(res => id = window.setTimeout(() => res(), timeout)),
```
```suggestion
    new Promise(res => {
      id = setTimeout(() => res(), timeout);
    }),
```
아예 줄바꿈을 해주는게 가독성에 더 좋을 것 같아요!

JunilHwang commented on src/endpoint/RestApi.js:
```diff
@@ -0,0 +1,45 @@
+    clearTimeout(id);
+    const contents = await race.json();
+
+    if (race.status === 500) new Error(contents.message);
+    if (race.status === 404) new Error('404');
```
여기서 throw를 던져주면 좋을 것 같아요! new Error만 하면 사실 별 의미가 없답니다

JunilHwang commented on src/endpoint/RestApi.js:
```diff
@@ -0,0 +1,45 @@
+
+  if (race instanceof Response) {
+    clearTimeout(id);
+    const contents = await race.json();
+
+    if (race.status === 500) new Error(contents.message);
+    if (race.status === 404) new Error('404');
+    return contents;
```
```suggestion
    if (race.status === 500) throw new Error(contents.message);
    if (race.status === 404) throw new Error('404');
    return race.json();
```
코드 순서는 이렇게 해도 좋을 것 같아요!

JunilHwang commented on src/endpoint/RestApi.js:
```diff
@@ -0,0 +1,45 @@
+    return contents;
+  }
+
+  else new Error('timeout');
```
```suggestion
  throw new Error('timeout');
```
마찬가지로 바로 throw를 해주세요!

JunilHwang commented on src/endpoint/service.js:
```diff
@@ -0,0 +1,113 @@
+  putUserTodoItemPriority,
+} from './api/user.js';
+
+export const createUser = async ({ name }) => {
+  const result = await postUser({ name });
+  // 에러 처리
+  return result;
+};
+
+export const removeUser = async ({ userId }) => {
+  const result = await deleteUser({ userId });
+  // 에러처리
+  return result;
+};
+
+export const readUserList = async () => {
+   const result = await getUserList();
+   // 에러 처리
+   return result;
+};
```
```suggestion
export const createUser = ({ name }) => postUser({ name });
export const removeUser = ({ userId }) => deleteUser({ userId });
export const readUserList = () => etUserList();
```
이렇게 표현할 수 있습니다!

JunilHwang commented on src/endpoint/service.js:
```diff
@@ -0,0 +1,113 @@
+  }
+  if (result.message === ERROR.DELETE_TODO_ITEM) {
+    throw new Error(ERROR.DELETE_TODO_ITEM);
+  }
+  return result;
+};
+
+export const updateUserTodoItem = async ({ userId, itemId, contents }) => {
+  const result = await putUserTodoItem({ userId, itemId, contents });
+  if (result.message === ERROR.UPDATE_TODO_ITEM) {
+    throw new Error(ERROR.UPDATE_TODO_ITEM);
+  }
+  if (result.message === ERROR.NO_USER3) {
+    throw new Error(ERROR.NO_USER3);
+  }
+
+  return result;
+};
+
+export const removeUserTodoItemsAll = async ({ userId }) => {
+  const result = await deleteUserTodoItemsAll({ userId });
+  if (result.success) {
+    result.message = SUCCESS.DELETE_TODO_ITEM_ALL;
+  }
+  if (result.message === ERROR.NO_USER3) {
+    throw new Error(ERROR.NO_USER3);
+  }
+  return result;
+};
+
+export const updateUserTodoItemPriority = async ({ userId, itemId, priority }) => {
+  const result = await putUserTodoItemPriority({ userId, itemId, priority });
+  if (result.message === ERROR.UPDATE_TODO_ITEM) {
+    throw new Error(ERROR.UPDATE_TODO_ITEM);
+  }
+  if (result.message === ERROR.NO_USER3) {
+    throw new Error(ERROR.NO_USER3);
+  }
+  return result;
+};
```
if로 하지말고 try로 해주세요

eyabc replied:
바꾸는데 하루종일 걸렸어요 ㅠㅠ

JunilHwang commented on src/store/index.js:
```diff
@@ -0,0 +1,100 @@
+};
+/* store 를 변경하는 함수는 set 으로 시작됩니다. */
+export const setter = {
+  async userList () {
+    try {
+      store.userList = await readUserList();
+    }
+    catch (err) {
+      alert(err.message);
+    }
+  },
```
```suggestion
  userList () {
      readUserList().then(userList => {
        store.userList = userList;
      }).catch(alert);
  },
```
async/await으로 값을 할당하면 proimse 형태로 들어갈 수 있기 때문에 이렇게 해주는게 더 좋을 것 같아요!

JunilHwang commented on src/store/index.js:
```diff
@@ -0,0 +1,100 @@
+      alert(err.message);
+    }
+  },
+  async user (userId) {
+    const newUserId = userId ? userId : store.userList[0]._id;
+    try {
+      const user = await readUser({ userId: newUserId });
+      store.user = user.message ? undefined : user;
+    }
+    catch (err) {
+      if (err.message === ERROR.NO_USER2) {
+        alert(err);
+        await initStore();
+      }
+    }
+    observer.render('user');
+  },
+  async userItems (userId) {
+    try {
+      store.user.todoList = await readUserTodoItems({ userId });
+    }
+    catch (err) {
+      if (err.message === ERROR.NO_USER2) {
+        alert(err.message);
+        await initStore();
+      }
+    }
+    finally {
+      observer.render('userItems');
+    }
+  },
```
앞선 리뷰와 동일합니다.

JunilHwang commented on src/store/index.js:
```diff
@@ -0,0 +1,100 @@
+    }
+  },
+  itemMode (itemId, mode) {
+    const item = store?.user?.todoList.find(v => v._id === itemId);
+    item.mode = mode;
```
item이  undefined일 가능성이 있기 때문에 이 부분은 사실 불안정한 코드입니다.

JunilHwang commented on src/store/index.js:
```diff
@@ -0,0 +1,100 @@
+    item.mode = mode;
+  },
+  userItem(itemId, newItem) {
+    let idx;
+    store.user.todoList.find((v, k) => {
+      if (v._id === itemId) {
+        idx = k;
+        return;
+      }
+    });
```
findIndex method를 사용해보세요~

JunilHwang reviewed (commented):
대체로 컴포넌트는 잘 설계해주신 것 같아요~
다만 promise/async/await/try/catch 등이 약간 미숙한 것 같습니다!
해당 부분만 보충하면 될 것 같아요 ㅎㅎ 👍

eyabc committed:
fix: userList 의 첫번째 유저가 존재하지 않을 때 무한 루프가 반복되는 버그 수정

eyabc committed:
refactor: 불필요한 문법 수정

eyabc committed:
refactor: export 변경

eyabc committed:
refactor: restAPI 에러 처리 수정

eyabc committed:
refactor: api 구조 변경

eyabc committed:
fix: user 생성시 해당 유저가 선택되지 않는 버그 수정

eyabc committed:
fix: user todo delete 버그 수정

eyabc committed:
refactor: 불필요한 async await 제거

wooooooood commented on README.md:
```diff
@@ -2,28 +2,72 @@
 
 이번 미션은 Ajax를 이용하여 TodoList를 발전시키는 미션입니다. 비동기 통신으로 데이터를 관리할 경우 자바스크립트의 비동기라는 특성을 더욱 깊이있게 이해할 수 있습니다
 
+## 하고싶은것
```
정리하신 내용이 인상적입니다!! 고민하신 흔적을 따라가기 좋은 것 같아요😀

wooooooood commented on src/store/index.js:
```diff
@@ -0,0 +1,75 @@
+  },
+};
+
+export const dispatch = {
```
이 부분은 사용하지 않는 코드인지 궁금합니다!

eyabc replied:
store  구조를 개선해 보고 싶었는데, 잘안됐어서 미루게 됐네요 ㅜㅜ 
깜박하고 지우지 않았어요 감사합니다

wooooooood commented on src/validator.js:
```diff
@@ -0,0 +1,7 @@
+
+  alert('User 의 이름은 최소 2글자 이상이어야 합니다.');
+  caller(validateUserName);
+};
```
소소하지만.. ㅎㅎ end of line이 빠져있는 소스 파일이 몇 개 있습니다!

eyabc replied:
저보다 꼼꼼하게 봐주셨네요 ㅋㅋ 잘 살펴보겠습니다!

wooooooood commented:
안녕하세요!! 리뷰가 늦어져 죄송합니다ㅠㅠ 수고 많으셨어요! 
js 초보라 제가 잘 모르는 문법이나 구조가 있어 코드 읽으면서 배운 것이 더 많은 시간이었습니다~^_^👍👍

wooooooood commented:
안녕하세요!! 늦은 리뷰를 남기게 되었습니다 ㅎㅎ 
이번 미션도 진행하시느라 수고 많으셨어요!
js 초보라 제가 잘 모르는 문법이나 구조가 있어 코드 읽으면서 배운 것이 더 많은 시간이었습니다~^_^👍👍

+) 헉 제가 Close를 잘못 눌러서 다시 Open했습니다.. 죄송합니당😂

eyabc committed:
Update index.js

pocojang commented on src/constants/filter.js:
```diff
@@ -0,0 +1,6 @@
+export const filter = {
```
1. 네이밍을 상수답게 바꾸시는게 어떨까요?!
2. 먼저 만든 상수를 재활용합시다!
3. `Object.freeze` 는 덤

```js
export const filter = {
  [CONST.ALL]: '전체보기',
  [CONST.ACTIVE]: '해야할 일',
  [CONST.COMPLETED]: '완료한 일'
};
```

eyabc replied:
레이어가 너무 많아서 저도 혼란스러웠는데 미션 3에서 api 를 가장 일찍 호출하는 부분은 store 와 연관이 깊더라고요 그래서 store 의 dispatch 라는 이름으로 api 호출하는 부분도 만들어 주었고. 함수형? 처럼 음... 설명은 잘 못하겠는데 2단계 정도 줄여보는 시도를 해보았습니다 ㅎㅎ

eyabc replied:
6a29578

미처 생각치 못한 부분 감사합니다!

pocojang commented on src/endpoint/HttpMethod.js:
```diff
@@ -0,0 +1,7 @@
+const GET = 'GET';
```
이런 친구들은 나중에 `Symbol()`로 더 견고하게 해보시는 것도 좋아요!

eyabc replied:
심볼에 대해서만 간단하게만 알고 있어서, 가이드를 더 주실 수 있을까요 ?

pocojang replied:
JS에서 고유한 값을 만들어낼 수 있는 방법이랍니다 ㅎㅎ!!

knae11 replied:
String 값을 상수로 사용하기 위해서 작성한 것 같은데 혹시 Object.freeze를 사용하지 않고 저렇게 한 이유를 알 수 있을까요...?👀

eyabc replied:
Object.freeze 에 대해서 제가 아는 선에서는,  해당 GET 이 Object 가 아니라 primitive 이기 때문에 사용하지 않았습니다! 다른의견이 있으시면 공유해주세요! 

제가 아직 깨달음이 부족해서 `const GET = 'GET'` 을 `Symbol()`  으로 사용하였을 때 무슨 장점이 있고 사용하지 않으면 예기치 못한 상황이라든지.. 

저의 상상인데, const 으로 문자열을 선언을 해도 해킹 이라든가 뭔가 문제가 있을까요? 

`const GET = 'GET';` 을 고유값으로 사용해야 하는 이유에 대해서 잘 모르겠습니다!

pocojang replied:
저 같은 경우 TypeScript를 주로 사용하기에 이러한 고민을 할 필요가 없는데요
@eyabc 님에게
상수를 관리할때 유니크한 상수에 대한 관점 정도로 소개를 한 것뿐이니 너무 개의치 않으셔도 됩니다! 🙇‍♂️

knae11 replied:
아! 그래서 freeze를 사용하지 않으신거군요!!ㅎㅎㅎ 잘 몰라서 질문한 것이었어요..!!감사합니당!

pocojang commented on src/endpoint/facade.js:
```diff
@@ -0,0 +1,158 @@
+import { ERROR } from '../constants/messageAPI.js';
+import { loadingWrapper, useMiddleWare } from '../utils.js';
+
+export const userCheck = async() => {
```
변수나 함수명을 작명하실때 동사를 `prefix`로 활용하시는게 좋아요!

eyabc replied:
3eda8b9 
감사합니다!

pocojang commented on src/endpoint/facade.js:
```diff
@@ -0,0 +1,158 @@
+  }
+};
+
+export const createUserFunc = async(name) => {
```
JS에서 `()`가 붙으면 어차피 함수를 의미하기에 함수를 콜백으로 넘기는 경우를 제외하고는
`func`라는 `suffix`를 사용할 경우가 많지 않습니다!

pocojang commented on src/endpoint/facade.js:
```diff
@@ -0,0 +1,158 @@
+import { getter, initStore, setter } from '../store/index.js';
+import { ERROR } from '../constants/messageAPI.js';
+import { loadingWrapper, useMiddleWare } from '../utils.js';
+
```
하나의 파일에서 여러가지 함수들을 모듈화할때 일부만 외부에서 사용한다면
일관적인 방법으로 export 하시는 걸 추천합니다

```js
export const func = () => // some code
const func2 = () => // some code
export const func3 = () => // some code
const func4 = () => // some code
```

```js
const func = () => // some code
const func2 = () => // some code
const func3 = () => // some code
const func4 = () => // some code

export {
  func,
  func3,
}
```

pocojang commented on src/store/index.js:
```diff
@@ -0,0 +1,70 @@
+    observer.render('userItems');
+  },
+  itemMode(itemId, mode) {
+    const item = store?.user?.todoList.find(v => v._id === itemId);
```
`find` 루프를 돌리면서 나오는 요소들을 `v`로 추상화하기보다는 제대로 모델링한다고 생각해보시는게 어떨까요?

pocojang commented on src/utils.js:
```diff
@@ -0,0 +1,35 @@
+  $loading && ($loading.style.display = display);
+};
+
+export const loadingWrapper = (handler) => {
```
동기적으로 작동될 수 있는 방법을  찾아보시면 더 좋은 함수가 되지 않을까 싶네요!

pocojang commented on src/utils.js:
```diff
@@ -0,0 +1,35 @@
+  showLoading(false);
+};
+
+export const useMiddleWare = async (middleware, handler) => {
```
미들웨어 역할을 적절하게 수행한다고 보기에는 조금 애매한 것 같아요!

1. 유저를 불러오는 부분에서만 사용되고 있다 => 차라리 순수 함수로 변경된다면 무의미해질 수 있는 함수
2. 미들웨어에서 처리할 수 있는 핸들러가 1개 밖에 없어 확장성이 부족하다.
=> 콜백 함수를 `Promise` 를 반환하는 함수들을 콜백으로 실행해주는 역할이 전부라 애매한 상황

pocojang commented on src/validator.js:
```diff
@@ -0,0 +1,7 @@
+export const validateUserName = async (name, caller) => {
+  if (name.length > 1)
```
one line 함수가 아니라면 중괄호로 감싸는게 더 좋지 않을까 싶네요!

eyabc replied:
감사합니다!
b038a14

pocojang commented on src/template/Todo/TodoLabel.js:
```diff
@@ -0,0 +1,47 @@
+// TODO refactor
+import { createDOM } from '../../utils.js';
+
```
1. 대문자 객체, 소문자 속성
2. 소문자 객체, 대문자 속성

2가지 케이스가 있는데 일관성을 잡아보는건 어떠실까요?

eyabc replied:
코드에서 어떤 부분을 가리키는지 잘모르겠는데, 좀 더 가이드를 주실 수 있을까요?

pocojang replied:
```js
// 케이스 1
const UPPER = {
  lower: '',
}

// 케이스 2
const lower = {
  UPPER: '',
}
```
이렇게 2가지 경우로 상수 값이 관리되고 있어서 규칙이 어떻게 될런지하는 질문입니다!

pocojang commented on src/store/index.js:
```diff
@@ -0,0 +1,70 @@
+
+export const initStore = async() => {
+  await readUserListFacade();
+  const userId = getter.userList()?.[0]._id || null;
```
`userList()`를 통해 억지로 꺼내오는 느낌이 있는데
역할에 맞는 함수를 차라리 하나 더 만드시는건 어떨까요?

eyabc replied:
50b1dfb
오호 감사합니다

pocojang commented on src/store/index.js:
```diff
@@ -0,0 +1,70 @@
+const store = {
+  userList: [],
+  user: undefined,
+  user1: {
```
user와 user1 차이는 무엇일까요?

eyabc replied:
실험적으로 써 놓은 건데 쓰지 않게 되었네요! 
d5ad502

넘나 꼼꼼하게 봐주셔서 감사합니다!

pocojang commented on src/endpoint/RestApi.js:
```diff
@@ -0,0 +1,45 @@
+    clearTimeout(id);
+    const contents = await race.json();
+    if (race.status === 404) throw new Error('404');
+    if (race.status === 500) throw new Error(contents.message);
```
여기서 에러 메세지를 담아주는데
에러메세지를 보여주는 곳에서는 지속적으로 분기문을 추가하셨는데
이유가 어떻게 되실까요?

eyabc replied:
분기문을 추가한 이유는, 
유저의 item 을 수정하거나 하는 요청을 보냈는데, 서버에 해당 user 가 존재하지 않는 경우는 message  로 에러에 대한 문자열을 보내주지만 응답 코드는 200 인 경우가 많아서 분기문을 지속적으로 추가해주었습니다. !

pocojang replied:
흠 그럼 `200`인데 불구하고 에러인 경우에 `catch` 절에 잡히나요??

eyabc replied:
그러게요? 제가 기억이 왜곡 되었나 봅니다 ㅋㅋ 다시한번 확인해보곘습니다 ㅠㅠ

pocojang reviewed (commented):
코드가 거대하고 계층이 여러 레이어로 나뉘다보니
이해하기가 어려웠네요 대단하시네요~

그만큼 이런 구조를 구성하고 작성하는데 얼마나 많이 고민하시고 시간을 쓰셨을지 상상이 안가네요!

eyabc committed:
refactor: 상수 모듈 변경
- https://github.com/next-step/js-todo-list-step2/pull/24#discussion_r515945209

eyabc committed:
Merge remote-tracking branch 'origin/master'

eyabc committed:
refactor: 변수나 함수명을 작명하실때 동사를 prefix로 활용
- https://github.com/next-step/js-todo-list-step2/pull/24#discussion_r515950947

eyabc committed:
refactor: 불필요한 프로퍼티 삭제
- https://github.com/next-step/js-todo-list-step2/pull/24#discussion_r515971840

eyabc committed:
refactor: 로직을 함수로 분리
- https://github.com/next-step/js-todo-list-step2/pull/24#discussion_r515971536

eyabc committed:
refactor: 네이밍 수정

eyabc committed:
refactor: if 문 중괄호
- https://github.com/next-step/js-todo-list-step2/pull/24#discussion_r515966737

pocojang commented on src/constants/messageAPI.js:
```diff
@@ -0,0 +1,13 @@
+export const ERROR = {
+  NO_USER: '해당 이름의 user가 없습니다.',  // used: createUserTodoItem
```
불필요한 주석 같아요!
저런건 에디터에서 충분히 확인 가능하지 않을까요?

eyabc replied:
87a375d

주석 out!!!

pocojang commented on src/constants/filter.js:
```diff
@@ -0,0 +1,12 @@
+  [COMPLETED]: '완료한 일'
+};
+
+Object.freeze(filter);
```
이런식으로 적용 가능하답니다!

```js
const ALL  = 'all';
const ACTIVE  = 'active';
const COMPLETED  = 'completed';

export const filter = Object.freeze({
  [ALL]: '전체보기',
  [ACTIVE]: '해야할 일',
  [COMPLETED]: '완료한 일'
});

// or
const filter = Object.freeze({
  [ALL]: '전체보기',
  [ACTIVE]: '해야할 일',
  [COMPLETED]: '완료한 일'
});

export {
  filter: Object.freeze(filter)
}
```

eyabc replied:
1b7ebbd
```javascript
export const filter = Object.freeze({
  [ALL]: '전체보기',
  [ACTIVE]: '해야할 일',
  [COMPLETED]: '완료한 일'
});

```
더 깔끔한 것 같네요!!

eyabc committed:
refactor: Object.freeze
- https://github.com/next-step/js-todo-list-step2/pull/24#discussion_r517341338

eyabc committed:
refactor: 주석 삭제
- https://github.com/next-step/js-todo-list-step2/pull/24#discussion_r517332425

pocojang commented on src/template/index.js:
```diff
@@ -0,0 +1,17 @@
+  showLoading(false);
+};
+
+
```
prettier를 사용하시는게 어떨까요?!
무의미한 개행이 지속적으로 보이네요!

pocojang commented on src/template/Todo/TodoContainer.js:
```diff
@@ -0,0 +1,22 @@
+const TodoContainer = () => {
+  const [getFilter, setFilter] = useState(CONST.ALL);
+
+  const dom = createDOM('section', { className: 'todoapp' });
```
그냥 `dom`은 아닌 것 같은데 이름을 달아주는 게 어떨까요?!

pocojang commented on src/endpoint/service.js:
```diff
@@ -0,0 +1,130 @@
+  deleteUserTodoItemsAll,
+  putUserTodoItemPriority,
+} from './api/user.js';
+
```
1. 반환이 없거나 반환을 전혀 활용하지 않는 함수에서도 무의미한 `return` 사용
2. 함수의 문맥상 `result` 보다는 `response` or `res` 등이 맞지 않을까요?! (Promise 객체를 반환하기때문)

knae11 commented on README.md:
```diff
@@ -2,28 +2,72 @@
+    - 2번을 먼저 해보고 1번을 시도해볼까? 
+    - 어떤게 나을까? 
+    
+- appendChild 말고 append 써보기
```
오~ 두 메서드의 차이에 대해서 알아보게 되었어요!!ㅎㅎㅎ

knae11 commented on src/endpoint/RestApi.js:
```diff
@@ -0,0 +1,45 @@
+  }
+
+  let id = -1;
+  const race = await Promise.race([
```
오 Promise.race라는 메서드 새롭게 배워갑니당!ㅎㅎㅎ

eyabc replied:
잘 쓰지 않는 메서드라고 하는데 코드스피츠 강의에서 주워먹어서 한번 써보았습니다!

knae11 commented on src/utils.js:
```diff
@@ -0,0 +1,35 @@
+  }
+};
+
+export const createDOM = (tagName, option = {}, dataset = undefined) => {
```
반복적으로 dom 엘리먼트를 만들기 위해 option 값까지 dom 엘리먼트를 만드는 함수를 구현해서 작성해 주셨군요!...👍 이건 궁금증인데요.. 위에 보다보면 render를 ` `(백틱) 안에 태그를 넣어서 하신 부분도 있더라구요! 어떤 경우에 createDom을 작성하고 어떤 경우에 백틱을 사용하신건지 기준을 물어봐도 될까요?? 일부러 createDom 사용해서 구현하려고 의도하신 것 같아 여쭤봅니다~!ㅎㅎㅎ

eyabc replied:
render 와 createDom 을 분리한 이유입니다.
- render 는 state 가 변경되었을 때 같이 다시 바뀌어야 하는 부분이고,
- createDOM 은 주로 이벤트를 등록하는 용도입니다.

만약에 두 부분이 합쳐져 있다면, 렌더할 때마다 이벤트를  등록해주어야 하기 때문에 분리하였습니다!

knae11 replied:
아 그런용도로 분리하신 거군요!!

knae11 reviewed (commented):
제가 아직 남의 코드를 읽고 이해할 수 있는 능력이 많이 안되서 궁금증 위주로 남깁니당!  모듈화가 익숙하지 않아서 어떻게 모듈로 나눌 수 있을까 고민하며 연습하고 있는데 엄청 작은 요소까지 모듈화하고 메서드로 만드신 것 같아 도움이 된 것 같아요! 👍

eyabc commented:
데브장님 살려주세요.. ㅠㅠ

pocojang commented:
> 데브장님 살려주세요.. ㅠㅠ

무엇을..  도와드리면 될까요?

eyabc committed:
refactor: api 변경
